Mathlib/Geometry/Manifold/LocalDiffeomorph.lean

```diff
@@ -0,0 +1,159 @@
+  {G : Type*} [TopologicalSpace G]
+  (I : ModelWithCorners 𝕜 E H) (J : ModelWithCorners 𝕜 F G)
+  (M : Type*) [TopologicalSpace M] [ChartedSpace H M]
+  (N : Type*) [TopologicalSpace N] [ChartedSpace G N] (n : ℕ∞)
+
+section LocalDiffeomorphAux
+/-- A "diffeomorphism on" `s` is a function `f : M → N` such that `f` restricts to a diffeomorphism
+`s → t` between open subsets of `M` and `N`, respectively.
+This is an auxiliary definition and should not be used outside of this file. -/
+structure LocalDiffeomorphAux extends LocalHomeomorph M N where
+  contMDiffOn_toFun : ContMDiffOn I J n toFun source
+  contMDiffOn_invFun : ContMDiffOn J I n invFun target
+
+/-- Coercion of a `LocalDiffeomorphAux` to function.
+Note that a `LocalDiffeomorphAux` is not `FunLike` (like `LocalHomeomorph`),
+as `toFun` doesn't determine `invFun` outside of `target`. -/
+instance : CoeFun (LocalDiffeomorphAux I J M N n) fun _ => M → N :=
+  ⟨fun Φ => Φ.toFun'⟩
+
+/-- A diffeomorphism is a local diffeomorphism. -/
+def Diffeomorph.toLocalDiffeomorphAux (h : Diffeomorph I J M N n) : LocalDiffeomorphAux I J M N n :=
+  {
+    toLocalHomeomorph := h.toHomeomorph.toLocalHomeomorph
+    contMDiffOn_toFun := fun x _ ↦ h.contMDiff_toFun x
+    contMDiffOn_invFun := fun _ _ ↦ h.symm.contMDiffWithinAt
+  }
+
+-- Add the very basic API we need.
+namespace LocalDiffeomorphAux
+variable (Φ : LocalDiffeomorphAux I J M N n) (hn : 1 ≤ n)
+protected theorem contMDiffOn : ContMDiffOn I J n Φ Φ.source :=
+  Φ.contMDiffOn_toFun
+
+protected theorem contMDiffOn_symm : ContMDiffOn J I n Φ.invFun Φ.target :=
+  Φ.contMDiffOn_invFun
+
+protected theorem mdifferentiableOn : MDifferentiableOn I J Φ Φ.source :=
+  (Φ.contMDiffOn).mdifferentiableOn hn
+
+protected theorem mdifferentiableOn_symm : MDifferentiableOn J I Φ.invFun Φ.target :=
+  (Φ.contMDiffOn_symm).mdifferentiableOn hn
+
+protected theorem mdifferentiableAt {x : M} (hx : x ∈ Φ.source) : MDifferentiableAt I J Φ x :=
+  (Φ.mdifferentiableOn hn x hx).mdifferentiableAt (Φ.open_source.mem_nhds hx)
+
+-- define symm just to make this easier to write?
+protected theorem mdifferentiableAt_symm {x : M} (hx : x ∈ Φ.source) :
+    MDifferentiableAt J I Φ.invFun (Φ x) :=
+  (Φ.mdifferentiableOn_symm hn (Φ x) (Φ.map_source hx)).mdifferentiableAt
+  (Φ.open_target.mem_nhds (Φ.map_source hx))
+
+/- We could add lots of additional API (following `Diffeomorph` and `LocalHomeomorph*), such as
+* further continuity and differentiability lemmas
+* refl, symm and trans instances; lemmas between them.
+As this declaration is meant for internal use only, we keep it simple. -/
+end LocalDiffeomorphAux
+end LocalDiffeomorphAux
+
+variable {M N}
+
+/-- `f : M → N` is called a **`C^n` local diffeomorphism at *x*** iff there exist
+  open sets `U ∋ x` and `V ∋ f x` and a diffeomorphism `Φ : U → V` such that `f = Φ` on `U`. -/
+def IsLocalDiffeomorphAt (f : M → N) (x : M) : Prop :=
+  ∃ Φ : LocalDiffeomorphAux I J M N n, x ∈ Φ.source ∧ EqOn f Φ Φ.source
```

Comment:
This doesn't seem like the most convenient definition of `IsLocalDiffeomorphAt`. For vector spaces we don't really give a definition, but in the statement of the inverse function theorem [HasStrictFDerivAt.to_localInverse](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Calculus/Inverse.html#HasStrictFDerivAt.to_localInverse) we use [HasStrictFDerivAt](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Calculus/FDeriv/Basic.html#HasStrictFDerivAt). Would it make sense to lift that definition to manifolds?